File: backend/src/models/campaign.py
from sqlalchemy import Column, DateTime, ForeignKey, Integer, Text, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from constants import CampaignStatus, CampaignType

from .audit import AuditMixin
from .base import Base, BaseMeta
from .campaign_customer_set import CampaignCustomerSet


class Campaign(Base, AuditMixin, metaclass=BaseMeta):
    __tablename__ = "campaign"

    id = Column(
        Text,
        nullable=False,
        primary_key=True,
        server_default=text("id_generator('campaign')"),
    )
    organization_id = Column(Text, ForeignKey("organization.id"), nullable=False)
    name = Column(Text, nullable=True)
    description = Column(Text, nullable=True)
    type = Column(Text, nullable=False, default=CampaignType.OUTBOUND)
    status = Column(Text, nullable=False, server_default=CampaignStatus.DRAFT.value)
    prompt = Column(Text, nullable=False)
    initial_message = Column(Text, nullable=False)
    max_duration = Column(Integer, nullable=False, default=0)
    max_retries = Column(Integer, nullable=False, default=0)
    end_date = Column(DateTime, nullable=True)

    # Overrides
    telephony_service_id = Column(Text, ForeignKey("telephony_service.id"), nullable=True)
    telephony_service_config = Column(JSONB, nullable=True)
    transcriber_id = Column(Text, ForeignKey("transcriber.id"),nullable=True)
    transcriber_config = Column(JSONB, nullable=True)
    agent_id = Column(Text, ForeignKey("agent.id"), nullable=True)
    agent_config = Column(JSONB, nullable=True)
    synthesizer_id = Column(Text, ForeignKey("synthesizer.id"), nullable=True)
    synthesizer_config = Column(JSONB, nullable=True)

    customer_sets = relationship(
        "CustomerSet",
        secondary=CampaignCustomerSet.__table__,
        lazy="selectin",
    )

File: frontend/types/campaign.ts
import { z } from "zod";
import { zodConfigJson } from "@/lib/utils";

export type CreateCampaign = {
    organization_id: string;
    name: string;
    description: string | null;
    type: string;
    status: string;
    prompt: string;
    initial_message: string;
    max_duration: number;
    max_retries: number;
    end_date: string | null;
    telephony_service_id?: string | null;
    telephony_service_config?: object | null;
    transcriber_id?: string | null;
    transcriber_config?: object | null;
    agent_id?: string | null;
    agent_config?: object | null;
    synthesizer_id?: string | null;
    synthesizer_config?: object | null;
    customer_sets: string[];
};

export type Campaign = CreateCampaign & {
    id: string;
    customer_sets: string[];
};

export const campaignFormSchema = z.object({
    name: z.string().min(2, { message: "Please enter a valid name" }),
    description: z.string().nullable(),
    type: z.string().min(2, { message: "Please enter a valid type" }),
    prompt: z.string().min(2, { message: "Please enter a valid prompt" }),
    initial_message: z.string(),
    // max_duration: z.number().int().positive(),
    // max_retries: z.number().int().positive(),
    telephony_service_id: z.string().nullable(),
    telephony_service_config: zodConfigJson,
    transcriber_id: z.string().nullable(),
    transcriber_config: zodConfigJson,
    agent_id: z.string().nullable(),
    agent_config: zodConfigJson,
    synthesizer_id: z.string().nullable(),
    synthesizer_config: zodConfigJson,
    customer_sets: z.array(z.string()),
});

export type CampaignFormValues = z.infer<typeof campaignFormSchema>;
